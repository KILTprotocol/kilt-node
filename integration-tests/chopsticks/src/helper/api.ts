/* eslint-disable @typescript-eslint/no-explicit-any */
import { ApiPromise } from '@polkadot/api'

export const xtokens = {
	parachainV2: (paraId: number) => (acc: any) => ({
		V1: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							network: 'Any',
							id: acc,
						},
					},
				],
			},
		},
	}),
	parachainV3: (paraId: number) => (acc: any) => ({
		V3: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							id: acc,
						},
					},
				],
			},
		},
	}),
	transfer:
		(token: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.xTokens.transfer(token, amount, dest(acc), weight),
}

export const balances = {
	transferAllowDeath: ({ api }: { api: ApiPromise }, dest: string, amount: string) =>
		api.tx.balances.transferAllowDeath(dest, amount),
}

export const switchPallet = {
	switchV3:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V3: {
					parents: 0,
					interior: {
						X1: {
							AccountId32: {
								id: acc,
							},
						},
					},
				},
			}),
	switchV4:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V4: {
					parents: 0,
					interior: {
						X1: [
							{
								AccountId32: {
									id: acc,
								},
							},
						],
					},
				},
			}),
	resume:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.resumeSwitchPair(),
	pause:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.pauseSwitchPair(),
}

export const xcmPallet = {
	parachainV2: (parents: number, paraId: number) => ({
		V2: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV3: (parents: number, paraId: any) => ({
		V3: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV4: (parents: number, paraId: any) => ({
		V4: {
			parents,
			interior: { X1: [{ Parachain: paraId }] },
		},
	}),
	limitedTeleportAssets:
		(token: any, amount: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedTeleportAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV2:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V2: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									network: 'Any',
									id: acc,
								},
							},
						},
					},
				},
				{
					V2: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV3:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV4:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV3:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV4:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV4:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V4: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V4: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV3:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V3: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V3: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	send:
		(destination: object) =>
		({ api }: { api: ApiPromise }, xcmMessage: object) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).send(destination, xcmMessage),
}

/**
 * Different pallets to submit tx
 */
export const tx = {
	xtokens,
	xcmPallet,
	switchPallet,
	balances,
}

/**
 * Query functions for different chains.
 * Native tokens are fetched via the system pallet, while other tokens are fetched via the tokens or assets pallet.
 *
 */
export const query = {
	balances: async ({ api }: { api: ApiPromise }, address: string) =>
		((await api.query.system.account(address)) as any).data.free.toBigInt(),
	foreignAssets:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.foreignAssets.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	fungibles:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.fungibles.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	tokens:
		(token: any) =>
		async ({ api }: { api: ApiPromise }, address: string) =>
			((await api.query.tokens.accounts(address, token)) as any).data.free.toBigInt(),
}
