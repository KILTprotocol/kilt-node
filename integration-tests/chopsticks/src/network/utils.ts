export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
