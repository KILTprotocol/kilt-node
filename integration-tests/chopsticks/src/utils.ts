import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
