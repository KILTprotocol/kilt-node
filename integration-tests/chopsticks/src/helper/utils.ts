import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
