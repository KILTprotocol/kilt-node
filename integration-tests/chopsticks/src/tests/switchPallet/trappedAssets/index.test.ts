import { describe, beforeEach, it, afterEach } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'
import type { KeyringPair } from '@polkadot/keyring/types'

import { createBlock, scheduleTx, setStorage } from '../../../network/utils.js'
import { hexAddress, keysAlice } from '../../../helper/utils.js'
import { testCases } from './config.js'
import type { Config } from '../../../network/types.js'
import { tx as txApi } from '../../../helper/api.js'
import { setupNetwork, shutDownNetwork } from '../../../network/utils.js'
import { skipTest } from '../../utils.js'

describe.skipIf(skipTest()).each(testCases)(
	'Reclaim trapped assets',
	{ timeout: 30_000 },
	async ({ account, query, txContext, config }) => {
		let senderContext: Config
		let receiverContext: Config
		let relayContext: Config
		let senderAccount: KeyringPair
		const { desc, network, storage } = config

		// Create the network context
		beforeEach(async () => {
			const { parachains, relay } = network

			const { parachainContexts, relayChainContext } = await setupNetwork(relay, parachains)
			const [senderChainContext, receiverChainContext] = parachainContexts

			relayContext = relayChainContext
			senderContext = senderChainContext
			receiverContext = receiverChainContext

			const { receiverStorage, senderStorage, relayStorage } = storage
			await setStorage(senderContext, senderStorage)
			await setStorage(receiverContext, receiverStorage)
			await setStorage(relayContext, relayStorage)

			senderAccount = account
		}, 20_000)

		// Shut down the network
		afterEach(async () => {
			try {
				await shutDownNetwork([receiverContext, senderContext, relayContext])
			} catch (error) {
				if (!(error instanceof TypeError)) {
					console.error(error)
				}
			}
		})

		it(desc, { timeout: 10_000, retry: 3 }, async ({ expect }) => {
			const { checkSystemEvents, checkEvents } = withExpect(expect)

			const { tx, balanceToTransfer, events, reclaimTx, getXcmMessage, senderLocation } = txContext

			// initial checks
			const balanceBeforeTx = await query.receiver(receiverContext, hexAddress(senderAccount.address))
			const balanceBeforeTxSender = await query.sender(senderContext, hexAddress(senderAccount.address))
			expect(balanceBeforeTx).toBe(BigInt(0))
			expect(balanceBeforeTxSender).toBeGreaterThan(BigInt(0))
			const rawTx = tx(senderContext, balanceToTransfer.toString())

			const events1 = await sendTransaction(rawTx.signAsync(senderAccount))

			// process tx
			await createBlock(senderContext)
			// process msg
			await createBlock(receiverContext)

			// check balance movement on sender chain.
			const balanceAfterTxSender = await query.sender(senderContext, hexAddress(senderAccount.address))
			expect(balanceAfterTxSender).toBe(balanceBeforeTxSender - balanceToTransfer)

			// Tx should fail on receiver
			const balanceAfterTx = await query.receiver(receiverContext, hexAddress(senderAccount.address))
			expect(balanceAfterTx).toBe(BigInt(0))

			// check events
			events.sender.map(
				async (pallet) =>
					await checkEvents(events1, pallet).toMatchSnapshot(
						`Switch eKILTs sender chain: ${JSON.stringify(pallet)}`
					)
			)

			await checkSystemEvents(receiverContext, 'polkadotXcm').toMatchSnapshot(
				'AssetsTrapped event on receiver chain'
			)

			// enable the switch pair again
			const resumeTx = txApi.switchPallet.resume()(receiverContext)
			scheduleTx(receiverContext, resumeTx)
			// process scheduled tx
			await createBlock(receiverContext)

			// create reclaim Tx
			const xcmMessage = getXcmMessage(balanceToTransfer.toString(), keysAlice.address)
			const rawReclaimTx = reclaimTx(senderContext, xcmMessage)

			// create reclaim message for relay chain
			const transactMessage = [
				{ UnpaidExecution: { weightLimit: 'Unlimited' } },
				{
					Transact: {
						originKind: 'SuperUser',
						requireWeightAtMost: { refTime: '1000000000', proofSize: '65527' },
						call: {
							encoded: rawReclaimTx.method.toHex(),
						},
					},
				},
			]

			// TODO: make relay reclaim tx configurable in the config
			const reclaimTxRelay = relayContext.api.tx.xcmPallet.send({ V3: senderLocation }, { V3: transactMessage })
			await sendTransaction(relayContext.api.tx.sudo.sudo(reclaimTxRelay).signAsync(keysAlice))
			// process tx
			await createBlock(relayContext)

			// check if the tx was successful
			await checkSystemEvents(relayContext, 'xcmPallet').toMatchSnapshot('reclaim xcm message on relay chain')

			// process and send message on sender chain.
			await createBlock(senderContext)

			// check if the tx was successful
			await checkSystemEvents(senderContext, 'polkadotXcm').toMatchSnapshot('reclaim xcm message on sender chain')
			// process message  receiver chain
			await createBlock(receiverContext)

			// check events
			events.receiver.map(
				async (pallet) =>
					await checkSystemEvents(receiverContext, pallet).toMatchSnapshot(
						`reclaim trapped assets receiver chain: ${JSON.stringify(pallet)}`
					)
			)
		})
	}
)
