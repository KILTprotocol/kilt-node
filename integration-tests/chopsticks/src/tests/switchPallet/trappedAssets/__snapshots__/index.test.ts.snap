// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reclaim trapped assets > V4 LIVE > AssetsTrapped event on receiver chain 1`] = `
[
  {
    "data": {
      "assets": {
        "V4": [
          {
            "fun": {
              "Fungible": 1000000000000000,
            },
            "id": {
              "interior": {
                "X2": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                  {
                    "AccountKey20": {
                      "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 2,
            },
          },
        ],
      },
      "hash_": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
    },
    "method": "AssetsTrapped",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > AssetsTrapped event on receiver chain 2`] = `
[
  {
    "data": {
      "assets": {
        "V4": [
          {
            "fun": {
              "Fungible": 1000000000000000,
            },
            "id": {
              "interior": {
                "X2": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                  {
                    "AccountKey20": {
                      "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 2,
            },
          },
        ],
      },
      "hash_": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
    },
    "method": "AssetsTrapped",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > Switch eKILTs sender chain: {"section":"polkadotXcm","method":"Sent"} 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": "All",
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > Switch eKILTs sender chain: {"section":"polkadotXcm","method":"Sent"} 2`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": "All",
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim trapped assets receiver chain: {"section":"assetSwitchPool1","method":"RemoteToLocalSwitchExecuted"} 1`] = `
[
  {
    "data": {
      "amount": 1000000000000000,
      "to": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "RemoteToLocalSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim trapped assets receiver chain: {"section":"assetSwitchPool1","method":"RemoteToLocalSwitchExecuted"} 2`] = `
[
  {
    "data": {
      "amount": 1000000000000000,
      "to": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "RemoteToLocalSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim trapped assets receiver chain: {"section":"polkadotXcm","method":"AssetsClaimed"} 1`] = `
[
  {
    "data": {
      "assets": {
        "V4": [
          {
            "fun": {
              "Fungible": 1000000000000000,
            },
            "id": {
              "interior": {
                "X2": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                  {
                    "AccountKey20": {
                      "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 2,
            },
          },
        ],
      },
      "hash_": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
    },
    "method": "AssetsClaimed",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim trapped assets receiver chain: {"section":"polkadotXcm","method":"AssetsClaimed"} 2`] = `
[
  {
    "data": {
      "assets": {
        "V4": [
          {
            "fun": {
              "Fungible": 1000000000000000,
            },
            "id": {
              "interior": {
                "X2": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                  {
                    "AccountKey20": {
                      "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 2,
            },
          },
        ],
      },
      "hash_": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
    },
    "method": "AssetsClaimed",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim xcm message on relay chain 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 0,
      },
      "message": [
        {
          "UnpaidExecution": {
            "checkOrigin": null,
            "weightLimit": "Unlimited",
          },
        },
        {
          "Transact": {
            "call": {
              "encoded": "(hash)",
            },
            "originKind": "Superuser",
            "requireWeightAtMost": {
              "proofSize": "(rounded 66000)",
              "refTime": 1000000000,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": "Here",
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "xcmPallet",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim xcm message on relay chain 2`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 0,
      },
      "message": [
        {
          "UnpaidExecution": {
            "checkOrigin": null,
            "weightLimit": "Unlimited",
          },
        },
        {
          "Transact": {
            "call": {
              "encoded": "(hash)",
            },
            "originKind": "Superuser",
            "requireWeightAtMost": {
              "proofSize": "(rounded 66000)",
              "refTime": 1000000000,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": "Here",
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "xcmPallet",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim xcm message on sender chain 1`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
          ],
        },
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "ClaimAsset": {
            "assets": [
              {
                "fun": {
                  "Fungible": 1000000000000000,
                },
                "id": {
                  "interior": {
                    "X2": [
                      {
                        "GlobalConsensus": {
                          "Ethereum": {
                            "chainId": 1,
                          },
                        },
                      },
                      {
                        "AccountKey20": {
                          "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 2,
                },
              },
            ],
            "ticket": {
              "interior": {
                "X1": [
                  {
                    "GeneralIndex": 4,
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": "All",
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": "Here",
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Reclaim trapped assets > V4 LIVE > reclaim xcm message on sender chain 2`] = `
[
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
          ],
        },
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "ClaimAsset": {
            "assets": [
              {
                "fun": {
                  "Fungible": 1000000000000000,
                },
                "id": {
                  "interior": {
                    "X2": [
                      {
                        "GlobalConsensus": {
                          "Ethereum": {
                            "chainId": 1,
                          },
                        },
                      },
                      {
                        "AccountKey20": {
                          "key": "0x5d3d01fd6d2ad1169b17918eb4f153c6616288eb",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 2,
                },
              },
            ],
            "ticket": {
              "interior": {
                "X1": [
                  {
                    "GeneralIndex": 4,
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": "All",
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": "Here",
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;
